qiskit/transpiler/coupling.py

```diff
@@ -403,6 +398,44 @@ def largest_connected_component(self):
         """Return a set of qubits in the largest connected component."""
         return max(rx.weakly_connected_components(self.graph), key=len)
 
+    def components(self, respect_direction: bool = False) -> List["CouplingMap"]:
+        """Separate a CouplingMap into subgraph Coupling Maps for each connected component.
+
+        This method will return a list of :class:`~.CouplingMap` objects for each connected
+        component in this :class:`~.CouplingMap`. The data payload of each node in the
+        :attr:`~.CouplingMap.graph` attribute will contain the qubit number in the original
+        graph. This will enables mapping the qubit index in a component subgraph to
+        the original qubit in the combined :class:`~.CouplingMap`. For example::
+
+            from qiskit.transpiler import CouplingMap
+
+            cmap = CouplingMap([[0, 1], [1, 2], [2, 0], [3, 4], [4, 5], [5, 3]])
+            component_cmaps = cmap.get_component_subgraphs()
+            print(component_cmaps[1].graph[0])
```

Comment:
Alternately, it seems like we're going through a lot of effort (and users will have to have some extra handling code) to workaround the issue that `CouplingMap` has to be indexed at zero, and can't have holes. Is now the right time to revisit that (maybe by _always_ using the node attribute data to store the physical qubit index)? This would require a different breaking change to users of `CouplingMap.graph`, but might be easier to manage in the long term.

Reply:
Actually the data payload for nodes in `CouplingMap.graph` is not always `None` it depends on how the `CoupingMap` was created. If you use the `add_physical_qubit()` api (also implicitly via `add_edge()`) it sets the node's payload to the index of the qubit. I thought about using a mapping dict, but it's not as straightforward as what we do in `DAGCircuit` because we have multiple graphs to work with. It'd basically have to be something like:
```
{
    0: (0, 0),
    1: (1, 0),
    2: (0, 1),
}
```
or a list of mappings:
```
[
    {
        0: 0,
        2: 1,
    },
    {
       1: 0,
    },
]
```
to indicate which subgraph index this node moved to and what it's corresponding index is in that subgraph. When thinking about how this mapping would be used I thought it would just be easier to have the node tell us what it was originally than having to do the double lookup.

What I'm actually protecting against here is not that everything has to be zero indexed and contiguous in the `CouplingMap`, but it's that we can't easily preserve indices across new graph object creation. If we wanted to do that I'd need a lot of logic to basically reconstruct the holes exactly to reuse the indices across all the subgraphs which is complex and error prone (but possible) and also a lot slower. It was much easier to just do the mapping once in the outer scope and let rustworkx just give us whatever indices make sense.

Reply:
> Actually the data payload for nodes in CouplingMap.graph is not always None it depends on how the CoupingMap was created. If you use the add_physical_qubit() api (also implicitly via add_edge()) it sets the node's payload to the index of the qubit.

Maybe this is a bug in and of itself. I tried `CouplingMap.from_*` and `CouplingMap(edge_list)` neither of which generated indices on the graph nodes. 

Agree that having the data on the nodes is preferable to a mapping dict, but I think really only if can be applied consistently. Otherwise, we risk ending up in a state where node data can be any of {always `None`, the physical qubit index on an originating `CouplingMap`, the physical qubit index on this `CouplingMap`} depending on how the `CouplingMap` was created.

As an alternative, would a move from having the node index to the node data be the source of truth for the physical qubit index be an option? Looking through `CouplingMap` quickly, I don't immediately see anything that would make this impossible (though I'm sure I'm missing pieces, and there are performance concerns that would have to be evaluated). This wouldn't I think be a change in existing behavior since for connected `CouplingMap`s, these would always be the same.

If that's not possible, I think returning a list of dicts that is already zipped with or is `zip`-able with the list of component graphs would also seem reasonable.

Reply:
I actually did a bit of a check and another place where it's not `None` currently is in the output from `Target.build_coupling_map()` (https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/transpiler/target.py#L839-L863 ). I can't remember why I did that, I assume I was thinking having the gates and error properties on the `CouplingMap`'s graph might be useful, especially for calculating shortest paths over error rates efficiently, so you could do something like `rx.dijkstra_shortest_paths(cmap.graph, 0, 2. weight_fn=lambda props: sum(x.error for x in props.values()))`. But nothing is really using it at least internally. So, I think we can safely say the node (and edge) payloads use is undefined at this point since we have 3 different use cases in tree depending on how it's created.

At the high level (which is wider scope than this PR) I think there are 2 ways to move forward here we can say the contents of these payloads on `CouplingMap.graph` continued to be undefined and their use is application specific and just document that explicitly at the class level. The other is to explicitly say that the node and edge payload is the canonical index for the node or edge on the backend. In general this will always be the rustworkx graph's index (e.g `cmap.graph[3] == 3`), the exception being those generated from this method which is doing this mapping. I'd really like to avoid as general practice having graph index != qubit index as I expect that will lead to more bugs unless you're explicitly handling it.

In general we'll want to use the index on the graph instead of the canonical index, we should only be relying on the node payload for mapping purposes like what is done here. The reason is if we're using the inner graph with any rustworkx functions (including `distance_matrix()`) the indices used are based on the graph index. I view this use as more of an edge case than something that will be used more generally outside of this. I'm not super worried about someone accidentally using the `CouplingMap`s output from this method and accidentally using index 0 as qubit 0 where it's really like qubit 2 globally because we're pretty explicit here in the docstring. The use case I had in mind for this is wrapping the `run()` method for layout and routing passes to basically split the circuit and coupling maps before and then run the pass for each subgraph and remap the output from the subgraph index to the global index.

That all being said if you would prefer this to be zippable list returns I can work with that, it's just a little extra mapping in the next PR to do it in with a dict instead of via the mapping interface the graph already gives us. The performance will likely be indistinguishable.

Reply:
> That all being said if you would prefer this to be zippable list returns I can work with that, it's just a little extra mapping in the next PR to do it in with a dict instead of via the mapping interface the graph already gives us. The performance will likely be indistinguishable.

I'm good to go ahead with this. I think the discussion around what should live on the nodes of `coupling_map.graph` and how we handle non-zero-indexed `CouplingMap`s and holes is an important one, but let's discuss that in a separate issue.